docs/indexing.rst

```diff
@@ -0,0 +1,123 @@
+Indexing
+========
+
+.. currentmodule:: nvidia.dali
+
+DALI data nodes can be indexed and sliced using familiar Python syntax for array indexing.
+
+Scalar indexing
+~~~~~~~~~~~~~~~
+
+The simplest case is scalar indexing with a constant index::
+
+    raw_files = fn.readers.file(...)
+    third_bytes = raw_files[2]
+
+The snippet above extracts the third (index 2) byte from the files.
+
+.. note::
+    The batch dimension is implicit and cannot be indexed. In this example, the index 2
+    is broadcast to the whole batch. The section :ref:`Run-time indices` for per-sample indexing.
+
+Indexing from the end
+~~~~~~~~~~~~~~~~~~~~~
+
+Negative indices can be used to index the tensor starting from the end. The index of -1 denotes
+the last element::
+
+    last_bytes = raw_files[-1]
+    penultimate_bytes = raw_files[-2]
+
+
+Run-time indices
+~~~~~~~~~~~~~~~~
+
+Indexing by constant is often insufficient. With DALI, you can use a result of other computations
+to access tensor elements::
+
+    length = fn.shapes(raw_files)[0]
+    index = fn.cast(fn.random.uniform(range=(0, 1)) * length, dtype=dali.types.INT64)
+    random_byte = raw_files[index]
+
+Here, a byte at random index will be extracted from each sample. ``index`` is a data node which
+represents a batch of scalar values, one per sample. Each of these values is used as the index
+for the respective sample in ``raw_files``.
+
+.. note::
+    The index must be a result of a CPU operator.
+
+
+Ranges
+~~~~~~
+
+To extract multiple values (or slices), the Python list slicing systax can be used::
+
+    header = raw_files[:16]
+
+If the start of the slice is omitted, the slice starts at 0. If the end is omitted, the slice
+ends at the end of given axis. Negative indices can be used for both start and end of the slice.
+Either end can be a constant, a run-time value (a DataNode) or can be skipped.
+
+Multidimensional indexing
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The index can contain multiple axes, separated by comma. If an index is a scalar, the corresponding
```

Comment:
I'm thinking how easy one can confuse scalars and constants here. Maybe we can find some better wording? Even writing `non-range` might help.

Also, the "dimension removal" can be bit misleading imo, maybe we should add a sentence, that the data is there, we just don't keep the 1-element-sized dummy axis?

Reply:
On the other hand that is how python works as well.

Reply:
@klecki We don't keep the axis. That''s the point.

Reply:
I'll reword it to "indexing" (scalar) and "slicing" (range).